bdk_electrum_example/src/main.rs

```diff
@@ -0,0 +1,169 @@
+            } else if all {
+                unused = false;
+                unspent = false
+            }
+            let mut spks: Box<dyn Iterator<Item = bdk_core::bitcoin::Script>> =
+                Box::new(core::iter::empty());
+            if unused {
+                spks = Box::new(spks.chain(txout_index.iter_unused().map(|(index, script)| {
+                    eprintln!("Checking if address at {:?} has been used", index);
+                    script.clone()
+                })));
+            }
+
+            if all {
+                spks = Box::new(spks.chain(txout_index.script_pubkeys().iter().map(
+                    |(index, script)| {
+                        eprintln!("scanning {:?}", index);
+                        script.clone()
+                    },
+                )));
+            }
+
+            if unspent {
+                spks = Box::new(spks.chain(tracker.utxos().map(|(_index, ftxout)| {
+                    eprintln!("checking if {} has been spent", ftxout.outpoint);
+                    ftxout.txout.script_pubkey
+                })));
+            }
+
+            let last_known_height = tracker
+                .chain()
+                .checkpoints()
+                .iter()
+                .last()
+                .map(|(&ht, _)| ht);
+
+            let mut scan_update = KeychainScan::default();
+
+            // Wallet scan returns a sparse chain that contains new All the BlockIds and Txids
+            // relevant to the wallet, along with keychain index update if required.
+            let new_sparsechain = client
+                .spk_txid_scan(spks, last_known_height)
+                .context("scanning the blockchain")?;
+
+            // Inserting everything from the new_sparsechain should be okay as duplicate
+            // data would be rejected at the time of update application.
+            for (height, hash) in new_sparsechain.checkpoints() {
+                let _ = scan_update.update.insert_checkpoint(BlockId {
+                    height: *height,
+                    hash: *hash,
+                })?;
+            }
+
+            // New Txids that we don't know the full data of.
+            let txid_changeset = tracker
+                .chain()
+                .determine_changeset(&new_sparsechain)?
+                .0
+                .txids;
+
+            let new_transactions =
+                client.batch_transaction_get(txid_changeset.iter().map(|(txid, _)| txid))?;
+
+            for (tx, index) in new_transactions
+                .into_iter()
+                .zip(txid_changeset.into_iter().map(|(_, index)| index))
+            {
+                scan_update.update.insert_tx(tx, index)?;
+            }
+
+            // Apply the full scan update
+            let changeset = tracker.determine_changeset(&scan_update)?;
+            db.append_changeset(&changeset)?;
+            tracker.apply_changeset(changeset);
+            debug!("sync completed!!")
```

Comment:
Let's put the duplicated code between sync and scan into a function.

Reply:
Good idea.. Was thinking exactly that..

Reply:
It seems like there is still a bit of duplicate code. Isn't it the case that everything after getting the sparse chain and applying the last active indexes will be the same?

Reply:
I think you can at least just return the `new_sparsechain` from the match arms and do the rest outside of the match statement.